[stable18]  Fix share transfer of single files and on the transfered node

juliusknorr opened the pull request:
Combined backport of #22116 #22648 including the new integration tests from #22761 and #22948

juliusknorr committed:
Transfer shares of the transferred root node

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Transfer shares if no path provided

Signed-off-by: Julius Härtl <jus@bitgrid.net>

danxuliu committed:
Check whether file exists or not after transferring ownership

The integration tests did not verify that the files were actually
transferred between the users, only that the files were downloadable.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Check share ownership after transferring file ownership

The integration tests did not verify that the shares were actually
transferred between the users (or that they were removed due to being
transferred to the sharee).

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for transferring ownership of reshares

Currently only transferring ownership of a reshare with a group to a
user in the group is possible.

When transferring ownership of a reshare with another user or with a
group to a user not in the group restoring the share fails (but the
command succeeds, it only fails for the specific files that are
reshares).

When transferring ownership of a path that is a reshare the command
fails (as when a specific path is provided the path tries to move the
file, it does not take into account reshares).

The added integration tests reflect the above behaviours.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

juliushaertl commented:
Oh, thanks for taking care already @danxuliu :heart:

danxuliu commented on lib/public/Files/Config/IUserMountCache.php:
```diff
@@ -117,4 +117,11 @@ public function remoteStorageMounts($storageId);
+	/**
+	 * Clear all entries from the in-memory cache
+	 *
+	 * @since 20.0.0
```
Should it be `@since 18.0.10`? I do not know how the `@since` tag is handled in backports.

MorrisJobke replied:
Leave it to 20. Otherwise we get into nightmares with 19.0.0 or something in between this and 19.0.4 to which is is ported.

danxuliu reviewed (commented):
> Oh, thanks for taking care already @danxuliu :heart:

My pleasure :-)

Regarding the sqlite failure it is not related to this pull request. Bisecting it in my system shows that it began in #21982. However, in Drone sometimes [passes](https://github.com/nextcloud/server/pull/22754) and sometimes [fails](https://github.com/nextcloud/server/pull/22746), so take that bisection with a pinch of salt ;-) Also it does not seem to be triggered when running only _apps/files_sharing/tests/SharedMountTest.php_, it seems to require running the full suite. So I do not know what is going on :shrug: But whatever it is, it is not related to this pull request :-P

MorrisJobke reviewed (approved):
Code looks good 👍

danxuliu committed:
Add integration test for transferring the path of a single file

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for transferring files of a user with a risky name

The files:transfer-ownership performs a sanitization of users with
"risky" display names (including characters like "\" or "/").

In order to allow (escaped) double quotes in the display name the
regular expression used in the "user XXX with displayname YYY exists"
step is not the "standard" one, "([^"]*)".

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests to check that only the given path is transferred

Until recently (it was fixed in ac2999a26a) when a path was transferred
other shares with the target user were removed, so a test was added to
ensure that it does not happen again.

Besides that a test to ensure that other files with the target user are
not transferred was added too (it did not fail before, but seemed
convenient to have that covered too :-) ).

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>
